File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.py
import sys
from collections import deque
input = sys.stdin.readline

N, M, R = map(int, input().split()) # 정점 수, 간선 수, 시작 정점
visited = [0] * (N+1)
G = [[] for _ in range(N+1)]
for _ in range(M):
    u, v = map(int, input().split())
    G[u].append(v)
    G[v].append(u)

dq = deque()
dq.append(R)
v_num = 1
visited[R] = v_num
v_num += 1
while dq:
    cur = dq.popleft()
    G[cur].sort()
    for nxt in G[cur]:
        if not visited[nxt]:
            visited[nxt] = v_num
            v_num += 1
            dq.append(nxt)

for i in visited[1:]:
    print(i)

File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.java
import java.io.*;
import java.util.*;

public class Main {
  static int N;
  static int M;
  static int R;
  static ArrayList<Integer>[] G;
  static int[] visited;
  static int visiteNum = 1;

  public static void bfs(int start) {
    Deque<Integer> dq = new ArrayDeque<>();
    visited[start] = visiteNum++;
    dq.offerLast(start);

    while (!dq.isEmpty()) {
      int cur = dq.pollFirst();
      for (int nxt:G[cur]) {
        if (visited[nxt] == 0) {
          visited[nxt] = visiteNum++;
          dq.offerLast(nxt);
        }
      }
    }
  }

  public static void main(String[] args) throws Exception{
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st;
    st = new StringTokenizer(br.readLine());
    N = Integer.parseInt(st.nextToken());
    M = Integer.parseInt(st.nextToken());
    R = Integer.parseInt(st.nextToken());

    G = new ArrayList[N+1];
    for (int i = 1; i <= N; i++) {
      G[i] = new ArrayList<>();
    }
    visited = new int[N+1];
    for (int i = 0; i < M; i++) {
      st = new StringTokenizer(br.readLine());
      int u = Integer.parseInt(st.nextToken());
      int v = Integer.parseInt(st.nextToken());
      G[u].add(v);
      G[v].add(u);
    }
    for (int i = 1; i <= N; i++) {
      Collections.sort(G[i]);
    }

    bfs(R);
    StringBuilder sb = new StringBuilder();
    for (int i = 1; i <= N; i++) {
      sb.append(visited[i]).append('\n');
    }
    System.out.println(sb);
  }
}
